include/fiction/algorithms/physical_design/hexagonalization.hpp

```diff
@@ -84,227 +294,638 @@ template <typename CartLyt, typename HexLyt>
-        detail::to_hex<CartLyt, HexLyt>(tile<CartLyt>(layout_width - 1, layout_height - 1, 0), layout_height).y;
-    const auto hex_width = detail::to_hex<CartLyt, HexLyt>(tile<CartLyt>(layout_width - 1, 0, 0), layout_height).x;
-    const auto hex_depth = layout_depth;
+        // compute hexagonal layout dimensions based on Cartesian dimensions
+        const auto hex_height =
+            detail::to_hex<CartLyt, HexLyt>({layout_width - 1, layout_height - 1, 0}, layout_height).y;
+        const auto hex_width = detail::to_hex<CartLyt, HexLyt>({layout_width - 1, 0, 0}, layout_height).x;
+        const auto hex_depth = layout_depth;
 
-    // instantiate hexagonal layout
-    HexLyt hex_layout{{hex_width, hex_height, hex_depth}, row_clocking<HexLyt>()};
+        // create the initial hexagonal layout
+        HexLyt hex_layout{{hex_width, hex_height, hex_depth}, row_clocking<HexLyt>()};
 
-    hexagonalization_stats stats{};
-    // measure run time
-    {
-        const mockturtle::stopwatch stop{stats.time_total};
+        // initialize statistics for hexagonalization
+        hexagonalization_stats stats{};
 
-        // calculate offset
-        const auto offset = detail::get_offset<HexLyt, CartLyt>(lyt, layout_width, layout_height);
+        {  // start timing the hexagonalization process
+            const mockturtle::stopwatch stop{stats.time_total};
 
-        // inputs
-        lyt.foreach_pi(
-            [&lyt, &hex_layout, &offset, &layout_height](const auto& gate)
-            {
-                const auto    old_coord = lyt.get_tile(gate);
-                tile<CartLyt> hex{detail::to_hex<CartLyt, HexLyt>(old_coord, layout_height)};
-                hex.x -= static_cast<decltype(hex.x)>(offset);
-                hex_layout.create_pi(lyt.get_name(lyt.get_node(old_coord)), hex);
-            });
-
-        // iterate through cartesian layout diagonally
-        for (uint64_t k = 0; k < layout_width + layout_height - 1; ++k)
-        {
-            for (uint64_t x = 0; x < k + 1; ++x)
-            {
-                const auto y = k - x;
-                if (y < layout_height && x < layout_width)
+            // calculate horizontal offset for hexagonal layout
+            const auto offset =
+                detail::get_offset<HexLyt, CartLyt>(plyt, layout_width, layout_height, ps.input_mode, ps.output_mode);
+
+            // determine the top primary input coordinate
+            auto middle_pi = detail::to_hex<CartLyt, HexLyt>({0, 0}, layout_height);
+
+            // determine the bottom primary output coordinate
+            auto middle_po = detail::to_hex<CartLyt, HexLyt>({plyt.x(), plyt.y()}, layout_height);
+
+            // vectors to store primary inputs
+            std::vector<tile<HexLyt>> left_pis;
+            std::vector<tile<HexLyt>> right_pis;
+
+            // vectors to store primary outputs
+            std::vector<tile<HexLyt>> left_pos;
+            std::vector<tile<HexLyt>> right_pos;
```

Comment:
Are the sizes of these vectors known? If so, using `.reserve()` would provide a speed-up.